### PR TITLE
more notifications fixes and unread counter

### PR DESCRIPTION
**Pull request by gmethvin**

### CODE REVIEW

File: packrat/main.js [hunk 1]
```diff
@@ -73,6 +73,7 @@ logEvent.catchUp = function() {
 // ===== WebSocket handlers
 
 const notifyCallbacks = [];
+var notifications = [], lastNotifyRead;
```
Comment (2is10) [inline packrat/main.js]: I moved the `notifications` declaration up to a section near the top of the file, for all cached server data.

File: packrat/main.js [hunk 2]
```diff
@@ -124,6 +109,36 @@ const socketHandlers = {
...
+    notifications.concat(data).forEach(function (n) {
+      idToNotif[n.id] = n;
+    });
+    notifications = [];
+    for (id in idToNotif) {
+      notifications.push(idToNotif[id]);
+    }
+    notifications.sort(function (a, b) {
+      var atime = new Date(a.time).getTime(), btime = new Date(b.time).getTime();
+      return btime - atime;
```
Comment (2is10) [inline packrat/main.js]: JS type coercion allows us to shorten this to: `return new Date(b.time) - new Date(a.time)`

File: packrat/main.js (hunk 1 above)
Comment (gmethvin) [inline packrat/main.js]: oh thanks; it looks like the change got lost in the merge

File: packrat/scripts/notices.js [hunk 3]
```diff
@@ -10,41 +10,44 @@ var MAX_NOTIFICATIONS = 100; // maximum number of notifications
...
-          });
-        }
+      getRenderedNotices(notices, $notifyPane, function () {
+        $notifyPane.scroll(function() {
+          var scrollBottom = $(this).scrollTop() + $(this).height();
+          var scrollHeight = $(this).prop("scrollHeight");
+          if (scrollHeight - scrollBottom < SCROLL_DISTANCE && number < MAX_NOTIFICATIONS) {
+            api.port.emit("notifications", number + 10, function (notices) {
+              getRenderedNotices(notices, $notifyPane, function () {
+                number += 10;
```
Comment (2is10) [inline packrat/scripts/notices.js]: how about `number = notices.length`, to assume less

File: packrat/scripts/notices.js [hunk 4]
```diff
@@ -10,41 +10,44 @@ var MAX_NOTIFICATIONS = 100; // maximum number of notifications
...
     if (~NOTICE_TYPES.indexOf(notice.category)) {
       render("html/metro/notice_" + notice.category + ".html", $.extend({
         formatMessage: getSnippetFormatter,
         formatLocalDate: getLocalDateFormatter,
         formatIsoDate: getIsoDateFormatter
       }, notice), function (html) {
-        renderedNotices.push(html);
+        renderedNotices[i] = html;
+        if (++done == notices.length) {
+          $notifyPane.html(renderedNotices);
```
Comment (2is10) [inline packrat/scripts/notices.js:44]: The two `.html` calls don't reset the scroll position? If not, wow, browsers have really advanced. If so, we should figure out a way to leave the DOM of the existing notifications in place, and just apply a diff.

**Comment by 2is10** [comment]
very nice!  feel free to merge.

File: packrat/scripts/notices.js (hunk 4 above)
Comment (gmethvin) [inline packrat/scripts/notices.js:44]: The scroll position doesn't change as long as you only modify stuff inside the div that scrolls. The reason I re-render all the things is that I think some of the previous notifications could potentially have changed.
